CMakeLists.txt

```diff
@@ -4,172 +4,127 @@ project(herb_description)
-        DEPENDS ${COMMAND_XACRO_POSTPROCESS} "${input_path}"
-        COMMAND ${COMMAND_XACRO_POSTPROCESS} --swap-dae
-                --name="${xacro_name}" --package="${PROJECT_NAME}"
-                --collision_meshes=True "${input_path}" "${output_path}"
-    )
-endmacro()
-
-# Create a standalone WAM URDF model.
-xacro_expand("${PROJECT_SOURCE_DIR}/config/wam_params.urdf.xacro"
-             "wam_params.urdf"
-)
-postprocess_urdf("${PROJECT_SOURCE_DIR}/robots/WAM_URDF.URDF"
-                 "wam_params.urdf"
-                 "wam_raw.urdf"
-)
-wrap_xacro("wam_raw.urdf"
-           "${URDF_OUTPUT_DIR}/wam.urdf.xacro"
-)
-
-xacro_expand("${URDF_OUTPUT_DIR}/wam_standalone.urdf.xacro"
-             "${URDF_OUTPUT_DIR}/wam.urdf"
-)
+option(BUILD_ALL_HERB_MODELS "Build all variations of the HERB models" OFF)
 
-add_custom_target(wam_urdf ALL
-    DEPENDS #"${URDF_OUTPUT_DIR}/wam.urdf"
-            "${URDF_OUTPUT_DIR}/wam.urdf.xacro"
-    COMMENT "Generating WAM URDF"
-    VERBATIM
-)
+macro(build_xacro infile outfile xacro_args)
+    # Call out to xacro to get dependencies
+    execute_process(COMMAND ${CATKIN_ENV} rosrun xacro xacro --deps ${infile}
+        ERROR_VARIABLE _xacro_err_ignore
+        OUTPUT_VARIABLE _xacro_deps_result
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-# Create a standalone BarrettHand (BH280) URDF model.
-xacro_expand("${PROJECT_SOURCE_DIR}/config/bh280_params.urdf.xacro"
-             "bh280_params.urdf"
-)
-postprocess_urdf("${PROJECT_SOURCE_DIR}/robots/BHD280_URDF.URDF"
-                 "bh280_params.urdf"
-                 "bh280_raw.urdf"
-)
-wrap_xacro("bh280_raw.urdf"
-           "${URDF_OUTPUT_DIR}/bh280.urdf.xacro"
-)
-
-xacro_expand("${URDF_OUTPUT_DIR}/bh280_standalone.urdf.xacro"
-             "${URDF_OUTPUT_DIR}/bh280.urdf"
-             # Additional dependencies.
-             "${URDF_OUTPUT_DIR}/bh280.urdf.xacro"
-)
-
-add_custom_target(bh280_urdf ALL
-    DEPENDS "${URDF_OUTPUT_DIR}/bh280.urdf"
-            "${URDF_OUTPUT_DIR}/bh280.urdf.xacro"
-    COMMENT "Generating BH280 URDF"
-    VERBATIM
-)
+    separate_arguments(_xacro_deps_result)
+    separate_arguments(xacro_args)
 
-# Create a WAM + BH280 model.
-xacro_expand("${URDF_OUTPUT_DIR}/wam_bh280_standalone.urdf.xacro"
-             "${URDF_OUTPUT_DIR}/wam_bh280.urdf"
-             # Additional dependencies.
-             "${URDF_OUTPUT_DIR}/bh280.urdf.xacro"
-             "${URDF_OUTPUT_DIR}/wam.urdf.xacro"
-)
-add_custom_target(wam_bh280_urdf ALL
-    DEPENDS #"${URDF_OUTPUT_DIR}/wam_bh280.urdf"
-            "${URDF_OUTPUT_DIR}/wam_bh280_standalone.urdf.xacro"
-    COMMENT "Generating WAM+BH280 URDF"
-    VERBATIM
-)
+    # Process xacro into final output
+    add_custom_command(OUTPUT ${outfile}
+        COMMENT "Generating ${outfile}"
+        COMMAND ${CATKIN_ENV} rosrun xacro xacro
+        -o ${outfile} ${infile} ${_xacro_deps_result} ${xacro_args}
+        DEPENDS ${infile} ${_xacro_deps_result}
+        VERBATIM)
+endmacro(build_xacro)
```

Comment:
I _think_ we should quote the variables that are used as a single argument, i.e. `"${outfile}"` and `"${infile}"`, to prevent breakage if the paths contain spaces.